js/components/thumbnail.js

```diff
@@ -7,12 +7,107 @@ var React = require('react'),
     Utils = require('./utils');
 
 var Thumbnail = React.createClass({
+  getInitialState: function() {
+    this.positionY = 0;
+    this.positionX = -320;
+    this.imageWidth = 0;
+    this.imageHeight = 0;
+    this.thumbnailWidth = 80;
+    this.thumbnailHeight = 40;
+    return {};
+  },
+  componentDidMount: function() {
+    this.setThumbnailSizes();
+    this.setImageSizes();
+    var yaw = this.props.vrViewingDirection.yaw;
+    var pitch = this.props.vrViewingDirection.pitch;
+    this.setCurrentViewVr(yaw, pitch);
+  },
+  componentWillReceiveProps: function(nextProps) {
+    if (this.props.vrViewingDirection !== nextProps.vrViewingDirection) {
+      var yaw = nextProps.vrViewingDirection.yaw;
+      var pitch = nextProps.vrViewingDirection.pitch;
+      this.setCurrentViewVr(yaw, pitch);
+    }
+  },
   shouldComponentUpdate: function(nextProps) {
-    return (nextProps.hoverPosition != this.props.hoverPosition);
+    var hoverPosition = nextProps.hoverPosition != this.props.hoverPosition;
+    var fullscreen  = nextProps.fullscreen != this.props.fullscreen && this.props.videoVr;
+    var vrViewingDirection = nextProps.vrViewingDirection != this.props.vrViewingDirection;
+    return (hoverPosition || fullscreen || vrViewingDirection);
+  },
+  componentDidUpdate: function(prevProps, prevState) {
+    var newThumbnailWidth = $("#oo-thumbnail").width();
+    var newThumbnailHeight = $("#oo-thumbnail").height();
+    if (newThumbnailWidth !== this.thumbnailWidth || newThumbnailHeight !== this.thumbnailHeight) {
+      this.thumbnailWidth = newThumbnailWidth;
+      this.thumbnailHeight = newThumbnailHeight;
+      var yaw = this.props.vrViewingDirection.yaw;
+      var pitch = this.props.vrViewingDirection.pitch;
+      this.setCurrentViewVr(yaw, pitch);
+    }
+  },
+
+  setThumbnailSizes: function() {
+    var thumbnailWidth = $("#oo-thumbnail").width();
+    var thumbnailHeight = $("#oo-thumbnail").height();
+    if (thumbnailWidth) {
+      this.thumbnailWidth = thumbnailWidth;
+    }
+    if (thumbnailHeight) {
+      this.thumbnailHeight = thumbnailHeight;
+    }
+  },
+
+  setImageSizes: function() {
+    var thumbnail = Utils.findThumbnail(this.props.thumbnails, this.props.hoverTime, this.props.duration, this.props.videoVr);
+    this.imageWidth = thumbnail.imageWidth;
+    this.imageHeight = thumbnail.imageHeight;
```

Comment:
(line 64-65) I think thumbnail size should not be dictated by the actual image's size. We do not want the thumbnail to blow up or shrink.

Reply:
Yes, we should not for 2d, but for 360 we have to do it.